Smart update of metric collection (#709)
PyTorchLightning/metrics: docs/source/pages/overview.rst

    have the same call signature. If this is not the case, input that should be
    given to different metrics can given as keyword arguments to the collection.

An additional advantage of using the ``MetricCollection`` object is that it will
automatically try to reduce the computations needed by finding groups of metrics

> that share the same underlying metric state. If such metrics are found only one

Reviewer: ```suggestion
that share the same underlying metric state. If such a group of metrics is found only one
```
can we also say that only the first os updated to reduce number one :)

> of them are actually updated and the updated state will be broadcastet to the rest

Reviewer 2: ```suggestion
of them is actually updated and the updated state will be broadcasted to the rest
```

> of the metrics. In the example above, this will lead to a 2x-3x lower computational

Reviewer: ```suggestion
of the metrics within the group. In the example above, this will lead to a 2x-3x lower computational
```

> cost compared to disabling this feature. See the `compute_groups` argument

Reviewer: ```suggestion
cost compared to disabling this feature. See the ``compute_groups`` argument
```

> cost compared to disabling this feature. See the `compute_groups` argument
> for more info.

Reviewer 2: ```suggestion
cost compared to disabling this feature. 
However, this speedup comes with a high cost upfront, where the state-groups have to be determined after the first update. This cost can be significant for very low steps (approx. up to 100) but still leads to an overall speedup for everything beyond that. In case the groups are known beforehand, these can also be set manually to avoid the cost of the dynamic search.
See the :param:`torchmetrics.MetricCollection.compute_groups` argument for more information on this topic.
```

Reviewer: > high cost upfront

how much high, this sounds like rather skip it...

Reviewer: I see so the overhead is fixed, so it may not pay off for very small scale training of about 100 batches but widely pays of for mid and large scale trainings

Reviewer: I think that this is the best place we want to show a graph which we can under as matplotlib figure
see: https://github.com/PyTorchLightning/pytorch-lightning/issues/5630

> 

Reviewer: maybe add some examples of such groups, so for classification, it is quite natural, but it may be hard for audio...

    have the same call signature. If this is not the case, input that should be
    given to different metrics can given as keyword arguments to the collection.

An additional advantage of using the ``MetricCollection`` object is that it will
automatically try to reduce the computations needed by finding groups of metrics
that share the same underlying metric state. If such a group of metrics is found only one
of them is actually updated and the updated state will be broadcasted to the rest
of the metrics within the group. In the example above, this will lead to a 2x-3x lower computational

> cost compared to disabling this feature. However, this speedup comes with a high cost upfront, where

Reviewer: ```suggestion
cost compared to disabling this feature. However, this speedup comes with a fixed cost upfront, where
```
the high sounds scary...

> the state-groups have to be determined after the first update. This cost can be significant for very

Reviewer: ```suggestion
the state-groups have to be determined after the first update. This overhead can be significantly higher then gains speed-up for very
```

> low steps (approx. up to 100) but still leads to an overall speedup for everything beyond that.

Reviewer: ```suggestion
a low number of steps (approx. up to 100) but still leads to an overall speedup for everything beyond that.
```

> In case the groups are known beforehand, these can also be set manually to avoid the cost of the

Reviewer: ```suggestion
In case the groups are known beforehand, these can also be set manually to avoid this extra cost of the
```
